x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/EsqlSession.java

```diff
@@ -125,14 +129,51 @@ public void execute(
         );
     }
 
+    /**
+     * Execute an analyzed plan. Most code should prefer calling {@link #execute} but
+     * this is public for testing. See {@link Phased} for the sequence of operations.
+     */
     public void executeAnalyzedPlan(
         EsqlQueryRequest request,
         BiConsumer<PhysicalPlan, ActionListener<Result>> runPhase,
         LogicalPlan analyzedPlan,
         ActionListener<Result> listener
     ) {
-        // TODO phased execution lands here.
-        runPhase.accept(logicalPlanToPhysicalPlan(analyzedPlan, request), listener);
+        LogicalPlan firstPhase = Phased.extractFirstPhase(analyzedPlan);
+        if (firstPhase == null) {
+            runPhase.accept(logicalPlanToPhysicalPlan(analyzedPlan, request), listener);
+        } else {
+            executePhased(new ArrayList<>(), analyzedPlan, request, firstPhase, runPhase, listener);
+        }
+    }
+
+    private void executePhased(
+        List<DriverProfile> profileAccumulator,
+        LogicalPlan mainPlan,
+        EsqlQueryRequest request,
+        LogicalPlan firstPhase,
+        BiConsumer<PhysicalPlan, ActionListener<Result>> runPhase,
+        ActionListener<Result> listener
+    ) {
+        PhysicalPlan physicalPlan = logicalPlanToPhysicalPlan(firstPhase, request);
+        runPhase.accept(physicalPlan, listener.delegateFailureAndWrap((next, result) -> {
+            try {
+                profileAccumulator.addAll(result.profiles());
+                LogicalPlan withPrevPhaseResults = Phased.applyResultsFromFirstPhase(mainPlan, physicalPlan.output(), result.pages());
+                LogicalPlan newNextPhase = Phased.extractFirstPhase(withPrevPhaseResults);
```

Comment:
I would name this as `newFirstPhase`.
From what I understand, these special Phased plans run (for now) in pairs of two plans. And `newNextPhase` should be basically the first phase of another INLINESTATS command in the same query. Is that right?

Reply:
Thinking some more about this, this Phased strategy opens the door to multi queries scenarios. Among others (for the future), there is one already that does something similar - metrics `rate` aggregate function where a mechanism of two internal aggregation functions linked together by a special data type for the purpose of sharing information between each other. @dnhatn I'm wondering if this Phased strategy would also work in case of `rate` function.

Reply:
:+1: